design/authorization.md

```diff
@@ -0,0 +1,85 @@
+
+Several operations can be exercised against a entity:
+* create a new entity
+* delete an entity
+* update entity's content
+
+Note that those operations combine themselves with the reference operations. For example to actually be able to update the content of an entity, user needs both permission to do the update AND to commit the change against the reference where the change will be stored
+
+### Other operations
+
+Some operations like performing garbage collection or changing access control do apply to the whole Nessie repository, and as such are control by their own specific set of permissions.
+
+
+## Server interface
+Although multiple server implementations may exist and conversely multiple access control interface may exist, this section aims to describe an interface to be used by the `JAX-RS` reference implementation of Nessie present under `servers/services`. The interface is composed of various methods for each operation to validate, which accept some access control context providing user identity and some arguments regarding the object to be accessed.
+
+More concretely, interface would look like this:
+
+    interface AccessChecker {
+      void canCreateReference(AccessContext context, NamedRef ref) throws AccessControlException;
+      void canDeleteReference(AccessContext context, NamedRef ref) throws AccessControlException;
+      [...]
+      void canReadValue(AccessContext context, Reference ref, ContentsKey key) throws AccessControlException;
+    }
+
+The `AccessContext` object passed as argument contains information regarding the overall context of the operation and will be created by the server itself:
+
+    interface AccessContext {
+      /**
+       * Provide a unique id for the operation being validated (for correlation purposes).
+       */
+      String operationId();
+
+      /**
+       * Provide the user identity.
+       */
+      Principal user();
+      
+      [...]
+    }
+
+## Server reference implementation
+
+An implementation of the `AccessChecker` interface could be written with the following characteristics:
+* Rules are written for either references or paths. Each rule contains a list of operations, and for each operation, a list of users (or group of users) and if the operation is granted or denied
+* To designate references or path, the use of wildcard is allowed. Rule ordering is based on how specific the match is. For example, `foo/bar/baz` matches both`foo/*` and `foo/bar/*`, but `foo/bar/*` is more specific, so rules under `foo/bar/*` would take precedence over `foo/*`
+* Rules would be stored under a special branch inside Nessie itself (meaning access control over rules would be done in the same manner as regular access control), under two different entities, one for the reference rules and one for the path rules (allowing for example to do an initial reference rule check at the edge).
```

Comment:
I like storing the rules in Nessie. eat our dogfood :smile: 

I wonder if rather than storing on a separate branch (presumably on a completely empty one to start) we store in a hidden directory. This has the effect of carrying the security permissions around with the whole db. So referencing a hash will have the rules at when the hash was committed and we could stage permissions with tables/views before merging as well. I suppose its not clear to me how we would revoke permissions to a particular table on a particular hash though....

seems like there are a handful of quirks/paradoxes listed above and part of the problem is the different histories (or lack of history) between the actual contents and the acls.